crates/sui-indexer-alt/src/queries/objects.rs

```diff
@@ -0,0 +1,246 @@
+                        hex::encode(package.to_vec())
+                    ));
+                }
+                ModuleFilter::ByModule(package, module) => {
+                    filter_conditions.push(format!(
+                        "package = '\\x{}'::bytea",
+                        hex::encode(package.to_vec()),
+                    ));
+                    filter_conditions.push(format!("module = '{}'", module));
+                }
+            },
+            TypeFilter::ByType(struct_tag) => {
+                filter_conditions.push(format!(
+                    "package = '\\x{}'::bytea",
+                    hex::encode(struct_tag.address.to_vec())
+                ));
+                filter_conditions.push(format!("module = '{:?}'", struct_tag.module.as_str()));
+                filter_conditions.push(format!("name = '{:?}'", struct_tag.name.as_str()));
+                filter_conditions.push(format!(
+                    "instantiation = '\\x{}'::bytea",
+                    hex::encode(bcs::to_bytes(&struct_tag.type_params).unwrap())
+                ));
+            }
+        }
+    }
+
+    filter_conditions.push(format!("cp_sequence_number <= {view_checkpoint_number}"));
+
+    if let Some(cursor) = cursor {
+        filter_conditions.push(format!(
+            "(cp_sequence_number < {} OR (cp_sequence_number = {} AND object_id > '\\x{}'::bytea))",
+            cursor.checkpoint,
+            cursor.checkpoint,
+            hex::encode(cursor.object_id)
+        ));
+    }
+    let filter_conditions_str = filter_conditions.join(" AND ");
+
+    let query = format!(
+        "
+        WITH filtered_rows AS (
+            SELECT
+                cp_sequence_number,
+                object_id
+            FROM
+                obj_info
+            WHERE
+                {filter_conditions_str}
+        )
+        SELECT
+            f.cp_sequence_number,
+            f.object_id
+        FROM
+            filtered_rows f
+        LEFT JOIN
+            obj_info o
+        ON
+            f.object_id = o.object_id
+            AND o.cp_sequence_number > f.cp_sequence_number
+            AND o.cp_sequence_number <= {view_checkpoint_number}
+        WHERE
+            o.object_id IS NULL
+        ORDER BY
+            f.cp_sequence_number DESC,
+            f.object_id ASC
+        LIMIT {limit};
+        ",
+    );
+
+    println!("{}", query);
+    let sql_query = sql_query(query);
+    let mut conn = db.connect().await?;
+    Ok(sql_query.load::<IdCheckpoint>(&mut conn).await?)
+}
+
+async fn query_latest_object_versions(
+    db: &Db,
+    objects: &[IdCheckpoint],
+) -> anyhow::Result<Vec<StoredObjVersion>> {
+    if objects.is_empty() {
+        return Ok(vec![]);
+    }
+    let conditions = objects
+        .iter()
+        .map(|o| {
+            format!(
+                "(object_id = '\\x{}'::bytea AND cp_sequence_number <= {})",
+                hex::encode(&o.object_id),
+                o.cp_sequence_number
```

Comment:
IIUC we only insert into `obj_info` when the ownership of an object changes, which means `obj_info` may not cover the latest checkpoint where the object was modified. And `o.cp_sequence_number` here is a cp sequence number we get from `obj_info` in the previous query. I don't get why we add "cp_sequence_number <= o.cp_sequence_number" as a condition. Should this be  "cp_sequence_number <= view_checkpoint_number" instead?

Reply:
You are totally right!